File: code/alien.py
import pygame
import math  # Import math for the sine function

class Alien(pygame.sprite.Sprite):
	def __init__(self,color,x,y):
		super().__init__()
		file_path = './graphics/' + color + '.png'
		self.image = pygame.image.load(file_path).convert_alpha()
		self.rect = self.image.get_rect(topleft = (x,y))
		self.bob_height = 1 
		self.bob_speed = 0.1 
		self.bob_offset = 0

		if color == 'red': self.value = 100
		elif color == 'green': self.value = 200
		else: self.value = 300

	def update(self,direction):
		self.rect.x += direction
		# Update bobbing effect
		self.bob_offset += self.bob_speed
		self.rect.y += math.sin(self.bob_offset) * self.bob_height  # Apply bobbing

class Extra(pygame.sprite.Sprite):
	def __init__(self,side,screen_width):
		super().__init__()
		self.image = pygame.image.load('./graphics/extra.png').convert_alpha()
		
		if side == 'right':
			x = screen_width + 50
			self.speed = - 3
		else:
			x = -50
			self.speed = 3

		self.rect = self.image.get_rect(topleft = (x,80))

	def update(self):
		self.rect.x += self.speed

File: space_station_native/SpaceStation/Sources/Models/Alien.swift
import SpriteKit

class Alien {
    enum AlienColor: String {
        case red, green, yellow

        var pointValue: Int {
            switch self {
            case .red: return 100
            case .green: return 300
            case .yellow: return 500
            }
        }
    }

    let sprite: SKSpriteNode
    let pointValue: Int
    private let color: AlienColor

    // Bobbing animation properties
    private let bobHeight: CGFloat = 1.0
    private let bobSpeed: CGFloat = 0.1
    private var bobOffset: CGFloat = 0.0

    init(color: AlienColor, position: CGPoint) {
        self.color = color
        self.pointValue = color.pointValue

        // Create alien sprite
        let texture = SKTexture(imageNamed: color.rawValue)
        sprite = SKSpriteNode(texture: texture)
        sprite.position = position
        sprite.zPosition = 5

        // Setup physics
        sprite.physicsBody = SKPhysicsBody(texture: texture, size: sprite.size)
        sprite.physicsBody?.isDynamic = true
        sprite.physicsBody?.categoryBitMask = PhysicsCategory.alien
        sprite.physicsBody?.contactTestBitMask = PhysicsCategory.playerLaser | PhysicsCategory.player | PhysicsCategory.block
        sprite.physicsBody?.collisionBitMask = PhysicsCategory.none
    }

    func update(direction: CGFloat) {
        sprite.position.x += direction

        // Apply bobbing effect
        bobOffset += bobSpeed
        let bobY = sprite.position.y + sin(bobOffset) * bobHeight
        sprite.position.y = bobY
    }
}
